src/NuGet.Services.AzureSearch/Owners2AzureSearch/OwnerDataClient.cs

```diff
@@ -0,0 +1,173 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.WindowsAzure.Storage;
+using Newtonsoft.Json;
+using NuGetGallery;
+
+namespace NuGet.Services.AzureSearch.Owners2AzureSearch
+{
+    public class OwnerDataClient : IOwnerDataClient
+    {
+        private static readonly JsonSerializer Serializer = new JsonSerializer();
+
+        private readonly ICloudBlobClient _cloudBlobClient;
+        private readonly IOptionsSnapshot<AzureSearchJobConfiguration> _options;
+        private readonly ILogger<VersionListDataClient> _logger;
+        private readonly Lazy<ICloudBlobContainer> _lazyContainer;
+
+        public OwnerDataClient(
+            ICloudBlobClient cloudBlobClient,
+            IOptionsSnapshot<AzureSearchJobConfiguration> options,
+            ILogger<VersionListDataClient> logger)
+        {
+            _cloudBlobClient = cloudBlobClient ?? throw new ArgumentNullException(nameof(cloudBlobClient));
+            _options = options ?? throw new ArgumentNullException(nameof(cloudBlobClient));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _lazyContainer = new Lazy<ICloudBlobContainer>(
+                () => _cloudBlobClient.GetContainerReference(_options.Value.StorageContainer));
+        }
+
+        private ICloudBlobContainer Container => _lazyContainer.Value;
+
+        public async Task<ResultAndAccessCondition<SortedDictionary<string, SortedSet<string>>>> ReadLatestIndexedAsync()
+        {
+            var blobName = GetLatestIndexedBlobName();
+            var blobReference = Container.GetBlobReference(blobName);
+
+            _logger.LogInformation("Reading the latest indexed owners from {BlobName}.", blobName);
+
+            var builder = new PackageIdToOwnersBuilder(_logger);
+            IAccessCondition accessCondition;
+            try
+            {
+                using (var stream = await blobReference.OpenReadAsync(AccessCondition.GenerateEmptyCondition()))
+                {
+                    accessCondition = AccessConditionWrapper.GenerateIfMatchCondition(blobReference.ETag);
+                    ReadStream(stream, builder.Add);
+                }
+            }
+            catch (StorageException ex) when (ex.RequestInformation.HttpStatusCode == (int)HttpStatusCode.NotFound)
+            {
+                accessCondition = AccessConditionWrapper.GenerateIfNotExistsCondition();
+                _logger.LogInformation("The blob {BlobName} does not exist.", blobName);
+            }
+
+            return new ResultAndAccessCondition<SortedDictionary<string, SortedSet<string>>>(
+                builder.GetResult(),
+                accessCondition);
+        }
+
+        public async Task UploadChangeHistoryAsync(IReadOnlyList<string> packageIds)
+        {
```

Comment:
Null check for packageIds?

Reply:
In general I avoid null checks all of the place in our methods. You'll notice 90% of our code doesn't null check method parameters. I think this is acceptable given our target consumer ... us!

It's a trade-off.

If we were releasing these .NET APIs publicly as packages and encouraging people to use them for their own projects, then I think we would want to be more rigorous about this.

We can maybe talk in the future within the team about adopting:
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#nullable-reference-types